### PR TITLE
[TOOLS-555] created dialog to save project configuration data

### PR DESCRIPTION
**Pull request by maykoone**

### CODE REVIEW

File: src/main/java/objective/taskboard/controller/ProjectController.java [hunk 1]
```diff
@@ -180,6 +188,48 @@ public void create(@RequestBody ProjectCreationData data) {
...
+
+        LocalDate startDate = DateTimeUtils.parseDate(data.startDate).toLocalDate();
+        LocalDate deliveryDate = DateTimeUtils.parseDate(data.deliveryDate).toLocalDate();
+        if (startDate.isAfter(deliveryDate))
+            return ResponseEntity.badRequest().body("{\"message\" : \"Invalid given dates\"}");
+
+        if (data.projectionTimespan != null && data.projectionTimespan < 0)
+            return ResponseEntity.badRequest().body("{\"message\" : \"Invalid projection timespan\"}");
+
+        configuration.get().setStartDate(startDate);
```
Comment (objective-solutions-user) [inline src/main/java/objective/taskboard/controller/ProjectController.java]: MAJOR Call "configuration.isPresent()" before accessing the value. [rule](http://10.44.1.250:9000/coding_rules#rule_key=squid%3AS3655)

File: src/main/java/objective/taskboard/controller/ProjectController.java [hunk 2]
```diff
@@ -180,6 +188,48 @@ public void create(@RequestBody ProjectCreationData data) {
...
+    }
+
+    @RequestMapping(value = "{projectKey}/configuration", method = RequestMethod.GET)
+    public ResponseEntity<ProjectConfigurationData> getProjectConfiguration(@PathVariable("projectKey") String projectKey) {
+        Optional<ProjectFilterConfiguration> configuration = projectRepository.getProjectByKey(projectKey);
+        if (!configuration.isPresent())
+            ResponseEntity.notFound().build();
+
+        ProjectConfigurationData data = new ProjectConfigurationData();
+        data.startDate = configuration.get().getStartDate() != null ? configuration.get().getStartDate().toString() : "";
```
Comment (objective-solutions-user) [inline src/main/java/objective/taskboard/controller/ProjectController.java]: MAJOR Call "configuration.isPresent()" before accessing the value. [rule](http://10.44.1.250:9000/coding_rules#rule_key=squid%3AS3655)

File: src/main/java/objective/taskboard/domain/ProjectFilterConfiguration.java [hunk 3]
```diff
@@ -49,17 +49,30 @@
     @Column
     private String projectKey;
     
+    @Column
+    private LocalDate startDate;
```
Comment (objective-solutions-user) [inline src/main/java/objective/taskboard/domain/ProjectFilterConfiguration.java]: MINOR Make this value-based field transient so it is not included in the serialization of this class. [rule](http://10.44.1.250:9000/coding_rules#rule_key=squid%3AS3437)

**Comment by objective-solutions-user** [comment]
SonarQube analysis reported 3 issues
* ![MAJOR][MAJOR] 2 major
* ![MINOR][MINOR] 1 minor

Watch the comments in this conversation to review them.

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: src/test/java/objective/taskboard/it/ProjectConfigurationDialog.java [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+        startDateInput.sendKeys("04/01/2018");
+        deliveryDateInput.sendKeys("02/01/2018");
+        projectionTimespanInput.sendKeys(Integer.valueOf(1).toString());
+        waitForClick(updateButton);
+        assertErrorMessage("Invalid given dates");
+        return this;
+    }
+
+    private void assertSuccessMessage(String message) {
+        waitUntil(new ExpectedCondition<Boolean>() {
```
Comment (jytakano) [inline src/test/java/objective/taskboard/it/ProjectConfigurationDialog.java]: Daria para usar o método `AbstractUiFragment.waitTextInElement` ?

File: src/test/java/objective/taskboard/it/ProjectConfigurationDialog.java [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+            public Boolean apply(WebDriver input) {
+                WebElement successMessage = dialog.findElement(By.cssSelector("#alertModalConfiguration .text"));
+                waitVisibilityOfElement(successMessage);
+                return message.equals(successMessage.getText()); 
+            }
+        });
+    }
+
+    private void assertErrorMessage(String message) {
+        waitUntil(new ExpectedCondition<Boolean>() {
```
Comment (jytakano) [inline src/test/java/objective/taskboard/it/ProjectConfigurationDialog.java]: Daria para usar o método `AbstractUiFragment.waitTextInElement` ?

File: src/test/java/objective/taskboard/it/ProjectConfigurationDialog.java [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+    }
+
+    public ProjectConfigurationDialog close() {
+        WebElement close = dialog.findElement(By.cssSelector(".button-close"));
+        waitForClick(close);
+        return this;
+    }
+
+    public ProjectConfigurationDialog updateConfiguration(String startDate, String deliveryDate, Integer projection) {
+        close();
```
Comment (jytakano) [inline src/test/java/objective/taskboard/it/ProjectConfigurationDialog.java]: Achei um pouco estranho os métodos começarem fechando e abrindo de novo.

File: src/main/resources/static/elements/taskboard/config-project-modal.html [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+                    projectKey: {
+                        type: String
+                    },
+                	projectConfiguration: {
+                        type: Object
+                    },
+                    showLoadingSpinner: {
+                        type: Boolean
+                    },
+                    dateFormaPlaceholder: {
```
Comment (jytakano) [inline src/main/resources/static/elements/taskboard/config-project-modal.html]: faltou o 't' do Format

File: src/main/resources/static/elements/taskboard/config-project-modal.html [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+                    console.log("update project configuration.");
+                    var self = this;
+
+                    if (!self._isValid())
+                        return;
+
+                    self.showLoading(true);
+
+                    var data = {
+                        startDate: self._getISODate(self.$.projectStartDate.$.input.value),
```
Comment (jytakano) [inline src/main/resources/static/elements/taskboard/config-project-modal.html]: Tem um esquema do Polymer que faz com que sempre que é alterado na tela, ele altera na property setada no value do componente, nesse caso, o "{{projectConfiguration.startDate}}". Aí não precisaria ficar pegando o value do componente.

File: src/main/java/objective/taskboard/controller/ProjectController.java [hunk 9]
```diff
@@ -180,6 +188,54 @@ public void create(@RequestBody ProjectCreationData data) {
...
+
+        LocalDate startDate = DateTimeUtils.parseDate(data.startDate).toLocalDate();
+        LocalDate deliveryDate = DateTimeUtils.parseDate(data.deliveryDate).toLocalDate();
+        if (startDate.isAfter(deliveryDate))
+            return ResponseEntity.badRequest().body("{\"message\" : \"End Date should be greater than Start Date\"}");
+
+        if (data.projectionTimespan != null && data.projectionTimespan < 0)
+            return ResponseEntity.badRequest().body("{\"message\" : \"Projection timespan should be a positive number\"}");
+
+        configuration.get().setStartDate(startDate);
```
Comment (objective-solutions-user) [inline src/main/java/objective/taskboard/controller/ProjectController.java]: MAJOR Call "configuration.isPresent()" before accessing the value. [rule](http://10.44.1.250:9000/coding_rules#rule_key=squid%3AS3655)

**Pull request merged**